### ISSUE TITLE
Strange behavior with UDF NewScope:0

### ISSUE THREAD

**Issue by SterLovesFood** [RPTools/maptool#2714]
**Describe the bug**
This is strange behavior that takes multiple steps to replicate and can have different results depending on the order code is arranged.

I've included an image that details what happens as clear as I can.

The campaign file I used to demonstrate this behavior is found at the bottom of this post.

Image:
![image](https://user-images.githubusercontent.com/45277765/121073215-dc016480-c797-11eb-8b36-cf7d168da484.png)

**Expected behavior**
In the image example, I think there is no scenario where 4 should be reaching chat output from inside an if() statement.
Even if the if statement begins with `[h,` in that example instead of `[r,`, 4 still reaches chat output.

**MapTool Info**
- Version: 1.8.5

**Desktop**
OS: Windows
Version 10.0.19042

**Campaign File**
https://www.dropbox.com/s/jkj78ovrc4vd7j3/BugTest.cmpgn?dl=0

**Comment by Irarara** [comment]
From what I can tell, this isn't actually an issue of scope. The issue is that the current behavior of UDFs is to return the value of `macro.return` if the output is empty regardless of whether `ignoreOutput` is `true` or not (See [here](https://github.com/RPTools/maptool/blob/3b2b94f46dd02320cfe56ce9e2c6bb2f564d09a2/src/main/java/net/rptools/maptool/client/functions/UserDefinedMacroFunctions.java#L152-L153)). 

Another simple test case to demonstrate this is to define two UDFs, `one` and `two`. Macro `one` should be `[h: two()]` and `two` should be `[h: macro.return = 1]`. Calling `[r: one()]` from the chat prints `1` to the chat, even though it should not print anything. 

Is there a reason for this behavior? It seems deliberate, but I'm not seeing the purpose behind it (implicit `ignoreOutput`?), and it's undocumented as far as I can tell (not mentioned on the wiki page for `defineFunction` at least).

**Comment by Phergus** [comment]
Ignore output is to suppress output from the commands *within* the UDF itself.  Not to suppress return values.

**Comment by Irarara** [comment]
Sure, but if `ignoreOutput` is `0`, then the output of the UDF should be the normal output of the macro being called, no? In this case, that would be nothing/an empty string.

Put another way, I would normally expect `[r: foo()]` (where the UDF was defined with `defineFunction("foo", "foo@Lib:Bar")`) to be functionally identical to `[r, macro("foo@Lib:Bar"): ""]`. If the macro in this case doesn't have any output but `macro.return` is set at some point, then that's not true. 

That all said, you _can_ trivially work around it by just making sure to set `macro.return` to `""` at the end of any UDF that should never output anything. It's still kind of confusing from a user perspective though IMO, and I'm not really seeing the benefit of the behavior. How many people even know UDFs work that way? I certainly didn't.

**Comment by Phergus** [comment]
This is how UDFs work and have worked since 1.5.1 and possibly before.

- If the UDF sets the value of `macro.return` and `ignoreOutput` is `0` then any output is the return value and `macro.return` has the value assigned.
- If the UDF sets the value of `macro.return` and `ignoreOutput` is `1` then the return value is the value assigned to `macro.return` and `macro.return` has the value assigned.

- If the UDF has output to chat, `macro.return` is **not** set and `ignoreOutput` is `0` then that output is the return value and `macro.return` is empty.
- If the UDF has output to chat, `macro.return` is **not** set  and `ignoreOutput` is `1` then  the return value and `macro.return` are empty.

![image](https://user-images.githubusercontent.com/34379254/121783748-9c5ac400-cb6d-11eb-8940-8f661440cac2.png)

I don't recall if there was discussion at the time UDFs were added as to what the return value should when `macro.return` is or isn't set but, as you said, I expect it was deliberate.

**Comment by Azhrei** [comment]
I added your table to [defineFunction()](https://wiki.rptools.info/index.php/defineFunction) on the wiki and cleaned up the page for `macro.return` as well.

**Comment by Phergus** [comment]
Good idea.  Thanks!

**Comment by SterLovesFood** [comment]
I just want to be clear as it feels like this discussion has gone in a somewhat different direction and I'm not sure if its just a related conversation or missing the bigger issue.

I'm entirely aware that ignoreOutput does not mean there is no output at all, but in the example code the UDF is placed within a situation where there just simply should not be any _chat_ output at all. Within a hidden if() statement:
`[h,if( ignore.Output(4) > 0 ),CODE:{};{}]`

And again, other irregularities with this only happening if the macro this executes from has no other chat output to seemingly override the issue. I'm not really sure if NewScope has anything to do with it or not, but that is how I became aware of this situation within my own code. It is entirely possible that the NoNewScope macro similarly is causing there to be no chat ouput when placed _before_ the if statement because even an empty r: macro is attempting to output a "" which has this same overriding behavior as any other output would have.


Just want to be clear in case this part is being normalized or overlooked

**Comment by Irarara** [comment]
Yeah, sorry, I did neglect to touch on that aspect of it. So, the reason you're seeing output is because of the whole deal with `macro.return`. `macro.return` is getting set (in `ignore.Output()`) , and since there is no other chat output it's being used as the return value in `normal.Function()`, since that's how UDFs (evidently) are designed to work. 

With your example, the reason it returns nothing if `no.Scope()` is called after `ignore.Output()` is because calling a new macro resets the value of `macro.return` (which makes sense, since `macro.return` is supposed to correspond with the last macro called).

**Comment by SterLovesFood** [comment]
Oh, I see...thanks for that explanation. It makes quite a bit more sense -- though still extremely unintuitive in this particular case.

**Comment by kwvanderlinde** [comment]
So for the example in the bug report, the issue is that `macro.return` values can "bubble up" from function calls. Is this deliberate/desirable behaviour, or is it just a natural consequence of using the same special variable to read return values as well as to write return values? Leaving aside backwards compatibility concerns, would it be considered better to have two separate variables or other mechanisms for interacting with return values? E.g., `macro.return` for returning values, `macro.lastReturned` for reading them. I'm just trying to get some discussion going rather than suggesting this is actually a good idea.

**Comment by kwvanderlinde** [comment]
> If the UDF sets the value of macro.return and ignoreOutput is 0 then any output is the return value and macro.return has the value assigned.

Just to be completely clear, there are actually two possible cases here: one where the output is empty and one where it is non-empty. So it works like this:
- If the UDF sets the value of macro.return and ignoreOutput is 0 and the output is empty, then the value of macro.return is the return value and macro.return has the value assigned.
- If the UDF sets the value of macro.return and ignoreOutput is 0 and the output is not empty, then the output is the return value and macro.return has the value assigned.

This extra case is necessary to explain what OP was seeing when adding a random "f" to the output. It's also basically what @Irarara explained above, I just wanted to make it clear how it factors into the case work. And, although I haven't looked at the implementation yet, I think ignoreOutput is kind of a red herring, as the actual case work depends on whether the output *happens to be* empty (ignoreOutput just changes the "happens to be" into a guarantee). What I mean is there are several ways of getting empty output so that `macro.return` is used:
- Of course, by setting ignoreOutput to 1, then: 
  ```
  [h: macro.return = "return value"]
  ```
- By hiding any output:
  ```
  [h: outputSomething()]
  [h: macro.return = "return value"]
  ```
- By using the return function:
  ```
  [h: return(0, "return value")]
  ```
- By more unpredictable means. E.g., if `outputSomething()` happens to not return anything, this will still use `macro.return` as the return value:
  ```
  [r: outputSomething()]
  [h: macro.return = "return value"]
  ```

**Comment by Irarara** [comment]
> So for the example in the bug report, the issue is that macro.return values can "bubble up" from function calls. Is this deliberate/desirable behaviour, or is it just a natural consequence of using the same special variable to read return values as well as to write return values? 

AFAIK, this is basically the purpose of `macro.return`. It's a means of getting a value out of a called function and pre-exists UDFs.

---

Since there's more discussion around this, I will say the for my part I would advocate for removing the behavior. If a person wants to use `macro.return` as the return value for a UDF, they would (or at least should) set `ignoreOutput` to `1`. Why would anyone ever want to rely on `macro.return` being returned on the condition of the rest of the output being empty? Trying to actually use this behavior seems like it's more likely to create issues than offer any tangible benefits.

Are there concerns about breaking existing macros/frameworks by changing it now? Given that it is/was undocumented behavior it seems unlikely that many people are actually relying on it (except maybe accidentally).
